### PR TITLE
Add Pagination on Daily Report Listing

### PR DESCRIPTION
**Pull request by queq1890**
# Story URL
https://trello.com/c/N0bwceQq/31-3-sprint2-daily-report-listing-subscription-pagination

# What 
- Add `$skip` `$first` variables to `userDailyReportsQuery`
- Add variables to `readQuery()` `writeQuery()` after create / delete reports
- Add pagination ( 10 items per page)

- Add `apollo-link-state` to store & access dailyReports count globally
- Add WebSocket settings for future subscription uses

### CODE REVIEW

**Comment by queq1890** [comment]
I will send you this pull request after #24 is merged

File: src/index.js [hunk 1]
```diff
@@ -4,30 +4,91 @@ import { Router } from 'react-router-dom';
...
+const link = split(
+  ({ query }) => {
+    const { kind, operation } = getMainDefinition(query);
+    return kind === 'OperationDefinition' && operation === 'subscription';
+  },
+  wsLink,
+  httpLink
+);
+
+const defaultState = {
```
Comment (queq1890) [inline src/index.js:52]: From line 52 to line 81, I'm initializing `apollo-link-state` based on the document
https://www.apollographql.com/docs/link/links/state.html

File: src/index.js [hunk 2]
```diff
@@ -4,30 +4,91 @@ import { Router } from 'react-router-dom';
...
+  process.env.NODE_ENV === 'production'
+    ? 'Set uri for production here'
+    : process.env.REACT_APP_GRAPHQL_URI;
+
+const httpLink = new HttpLink({
+  credentials: 'include',
+  uri: httpUri
+});
+
+const wsUri =
```
Comment (queq1890) [inline src/index.js:31]: From line 31 to line 50, I'm initializing Websocket settings for subscription.

File: src/containers/DailyReportPage/DailyReportContainer.js [hunk 3]
```diff
@@ -13,84 +12,146 @@ import {
...
-      if (isEmpty(data.userReports)) {
-        return (
-          <ContentWrapper>
-            <h2>No Daily Report is submitted yet!</h2>
-          </ContentWrapper>
-        );
-      }
+export const USER_DAILY_REPORTS_COUNT_QUERY = gql`
+  query USER_DAILY_REPORTS_COUNT_QUERY {
+    userDailyReportsCount @client {
```
Comment (queq1890) [inline src/containers/DailyReportPage/DailyReportContainer.js:43]: Query & Mutation with `@client` only interact with the cache and never get data from server side.
https://www.apollographql.com/docs/link/links/state.html

File: src/index.js [hunk 4]
```diff
@@ -4,30 +4,95 @@ import { Router } from 'react-router-dom';
...
+          ...previousState,
+          userDailyReportsCount: {
+            __typename: 'UserDailyReportsCount',
+            count,
+            isInitialized: true
+          }
+        };
+
+        cache.writeData({ query: USER_DAILY_REPORTS_COUNT_QUERY, data });
+        return null;
```
Comment (queq1890) [inline src/index.js:81]: returning `null` due to `apollo-link-state` specification 
https://github.com/apollographql/apollo-link-state/issues/160

File: .env.sample [hunk 5]
```diff
@@ -3,3 +3,4 @@
 
 SKIP_PREFLIGHT_CHECK=
 REACT_APP_GOOGLE_CLIENT_ID=
+REACT_APP_WS_URI=
```
Comment (queq1890) [inline .env.sample:6]: I added WebSocket endpoint to `.env`
`ws://......`

**Review by queq1890 — COMMENTED** [review]
I added `apollo-link-state` to access to current DailyReports count much easier, 
but pagination logics maybe be going to be changed lots after changing DailyReports order 
https://trello.com/c/qe5W8wK9

**Review by ggianglv — APPROVED** [review]
LGTM

File: src/containers/DailyReportPage/DailyReportContainer.js [hunk 6]
```diff
@@ -15,93 +14,152 @@ import {
...
+`;
 
-      if (error) return <p>Error: {error.message}</p>;
+class DailyReportContainer extends Component {
+  nextPage = data => {
+    const currentPage = Number(queryString.parse(this.props.location.search).page) || 1;
+    const { count } = data.userReports;
+    const nextPage = currentPage + 1;
+    if (currentPage <= count / itemsAmount) {
+      this.props.history.push(`/reports?page=${nextPage}`);
```
Comment (phuongthuan) [inline src/containers/DailyReportPage/DailyReportContainer.js]: I have created a custom history in utils. 
Import `history` from `utils` and we can use this to navigate.

File: src/containers/DailyReportPage/DailyReportContainer.js [hunk 7]
```diff
@@ -15,93 +14,152 @@ import {
...
-          <ContentWrapper>
-            <h2>No Daily Report is submitted yet!</h2>
-          </ContentWrapper>
-        );
-      }
+  prevPage = () => {
+    const currentPage = Number(queryString.parse(this.props.location.search).page) || 1;
+    const prevPage = currentPage - 1;
+    if (currentPage > 1) {
+      this.props.history.push(`/reports?page=${prevPage}`);
```
Comment (phuongthuan) [inline src/containers/DailyReportPage/DailyReportContainer.js]: Same above.

File: src/containers/DailyReportPage/index.js [hunk 8]
```diff
@@ -15,8 +15,8 @@ const DailyReportPage = ({ match }) => (
     </LeftContent>
     <RightContent>
       <Switch>
-        <Route exact path={match.url} component={DailyReportContainer} />
-        <Route path={`${match.path}/new`} component={CreateDailyReportContainer} />
+        <Route exact path={match.url} render={props => <DailyReportContainer {...props} />} />
+        <Route path={`${match.path}/new`} render={props => <CreateDailyReportContainer {...props} />}/>
```
Comment (phuongthuan) [inline src/containers/DailyReportPage/index.js]: By using `history` in `utils`, we don't need to pass `...props` to the component like this. 

Should be the same as before.
```
<Route exact path={match.url} component={DailyReportContainer} />
```

**Review by phuongthuan — COMMENTED** [review]
I have some comment.

File: src/containers/DailyReportPage/index.js (hunk 8 above)
Comment (queq1890) [reply src/containers/DailyReportPage/index.js]: @phuongthuan 
Thanks, however, I actually needed to pass `location` by `react-router` to access query string from children components in `DailyReportContainer`.

Like, 
location.seacrh
=> `page?=1`

So I'm gonna leave `DailyReportContainer` as it is and fix `CreateDailyReportContainer`.

File: src/containers/DailyReportPage/DailyReportContainer.js (hunk 7 above)
Comment (queq1890) [reply src/containers/DailyReportPage/DailyReportContainer.js]: @phuongthuan  Fixed both. Thank you for comments!

**Pull request merged**